### PR TITLE
support restore from S3 with BR

### PR DESCRIPTION
**Pull request by francis0407**
### What problem does this PR solve? 
This PR integrates BR with tidb-operator to support restore data from S3.

### Check List 

Tests 

 - Manual test (add detailed scripts or steps below)

Code changes

 - Has Go code change
 - Has documents change

Side effects

 - Breaking backward compatibility

Related changes

 - Need to update the documentation

### Does this PR introduce a user-facing change?:
 
 ```release-note

 ```

### CODE REVIEW

File: cmd/backup-manager/app/util/remote.go [hunk 1]
```diff
@@ -0,0 +1,249 @@
+// Copyright 2019 PingCAP, Inc.
```
Comment (gregwebs) [inline cmd/backup-manager/app/util/remote.go]: can we put this into an s3 module? At least the filename could be s3.

File: cmd/backup-manager/app/util/remote.go [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return blob.PrefixedBucket(bkt, qs.prefix), nil
+
+}
+
+// checkS3Config constructs s3Query parameters
+func checkS3Config(s3 *v1alpha1.S3StorageProvider, fakeRegion bool) *s3Query {
+	sqs := s3Query{}
```
Comment (gregwebs) [inline cmd/backup-manager/app/util/remote.go]: there is another amazon service called "sqs". Seeing this variable name confused me at first.

File: cmd/backup-manager/app/util/remote.go [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+	sqs.prefix = s3.Prefix
+	sqs.endpoint = s3.Endpoint
+	sqs.oriEndpoint = s3.Endpoint
+
+	if fakeRegion && sqs.region == "" {
+		sqs.region = "us-east-1"
+	}
+	sqs.prefix = strings.Trim(sqs.prefix, "/")
+	sqs.prefix += "/"
+	if sqs.endpoint != "" {
```
Comment (gregwebs) [inline cmd/backup-manager/app/util/remote.go]: can the endpoint begin with `s3://` ?

**Comment by francis0407** [comment]
I have manually tested this new feature. PTAL. cc @DanielZhangQD, @onlymellb

**Comment by onlymellb** [comment]
@francis0407 Can you resolve the conflict？

File: cmd/backup-manager/app/util/util.go [hunk 4]
```diff
@@ -111,13 +111,47 @@ func SetFlagsFromEnv(flags *pflag.FlagSet, prefix string) error {
...
+	if err != nil {
+		return nil, err
+	}
+	args = append(args, storageArgs...)
+	if (restore.Spec.Type == v1alpha1.BackupTypeDB || restore.Spec.Type == v1alpha1.BackupTypeTable) && config.DB != "" {
+		args = append(args, fmt.Sprintf("--db=%s", config.DB))
+	}
+	if restore.Spec.Type == v1alpha1.BackupTypeTable && config.Table != "" {
+		args = append(args, fmt.Sprintf("--table=%s", config.Table))
+	}
```
[the remark this answers is not in the document]
Comment (francis0407) [reply cmd/backup-manager/app/util/util.go]: If DB or Table is not set, the error has been thrown in `ValidateRestore`.

**Comment by francis0407** [comment]
Conflicts have been resolved. PTAL @onlymellb

File: cmd/backup-manager/app/restore/restore.go [hunk 5]
```diff
@@ -68,5 +69,8 @@ func constructBROptions(restore *v1alpha1.Restore) ([]string, error) {
 	if config.Checksum != nil {
 		args = append(args, fmt.Sprintf("--checksum=%t", *config.Checksum))
 	}
+	if config.RateLimit != nil {
```
Comment (DanielZhangQD) [inline cmd/backup-manager/app/restore/restore.go]: There should be an `online` option.

**Pull request merged**